File: src/observations.py
"""Observation CRUD with entity linking and embedding."""

import json

import aiosqlite

from src.config import Config
from src.embeddings import embed_texts, serialize_embedding
from src.models import (
    AddResult,
    ChunkResult,
    ObservationDetail,
    ObservationInput,
    utcnow,
)


async def add_observations(
    db: aiosqlite.Connection,
    observations: list[ObservationInput],
    config: Config,
) -> AddResult:
    """Batch-add observations. Auto-embeds content and links to entities."""
    result = AddResult()
    if not observations:
        return result

    # Batch embed all observation content
    texts = [obs.content for obs in observations]
    embeddings = await embed_texts(texts, config)

    now = utcnow()
    for obs, embedding in zip(observations, embeddings, strict=True):
        try:
            metadata_json = json.dumps(obs.metadata)
            cursor = await db.execute(
                """INSERT INTO observations
                   (content, metadata, confidence, provenance, created_at)
                   VALUES (?, ?, ?, ?, ?)""",
                (obs.content, metadata_json, obs.confidence, obs.provenance, now),
            )
            obs_id = cursor.lastrowid
            if obs_id is None:
                result.errors.append("Failed to insert observation (no ID returned)")
                continue

            # Insert embedding into vec0
            blob = serialize_embedding(embedding)
            await db.execute(
                "INSERT INTO observation_embeddings (observation_id, embedding) VALUES (?, ?)",
                (obs_id, blob),
            )

            # Link to entities
            for entity_name in obs.entity_names:
                ecursor = await db.execute("SELECT id FROM entities WHERE name = ?", (entity_name,))
                erow = await ecursor.fetchone()
                if erow is not None:
                    await db.execute(
                        """INSERT OR IGNORE INTO observation_entities
                           (observation_id, entity_id) VALUES (?, ?)""",
                        (obs_id, erow["id"]),
                    )
                else:
                    result.errors.append(
                        f"Entity '{entity_name}' not found, skipping link for observation"
                    )

            result.added += 1
        except Exception as e:
            result.errors.append(f"Observation: {e}")

    await db.commit()
    return result


async def read_observation(
    db: aiosqlite.Connection,
    observation_id: int,
) -> ObservationDetail | None:
    """Read full observation by ID, including linked entity names."""
    cursor = await db.execute(
        "SELECT id, content, metadata, created_at FROM observations WHERE id = ?",
        (observation_id,),
    )
    row = await cursor.fetchone()
    if row is None:
        return None

    # Fetch linked entity names
    ecursor = await db.execute(
        """SELECT e.name FROM entities e
           JOIN observation_entities oe ON e.id = oe.entity_id
           WHERE oe.observation_id = ?""",
        (observation_id,),
    )
    entity_rows = await ecursor.fetchall()

    return ObservationDetail(
        id=row["id"],
        content=row["content"],
        metadata=json.loads(row["metadata"]),
        entity_names=[r["name"] for r in entity_rows],
        created_at=row["created_at"],
    )


def _split_into_chunks(text: str, chunk_size: int, overlap: int) -> list[str]:
    """Split text at sentence boundaries near chunk_size."""
    if len(text) <= chunk_size:
        return [text]

    chunks: list[str] = []
    start = 0
    while start < len(text):
        end = start + chunk_size
        if end >= len(text):
            chunks.append(text[start:])
            break
        # Find sentence boundary near end
        boundary = -1
        for sep in [". ", ".\n", "\n\n", "\n", " "]:
            found = text.rfind(sep, start + chunk_size // 2, end + 1)
            if found != -1:
                boundary = found + len(sep)
                break
        if boundary == -1:
            boundary = end
        chunks.append(text[start:boundary])
        start = max(start + 1, boundary - overlap)
    return chunks


async def add_chunked_observation(
    db: aiosqlite.Connection,
    content: str,
    config: Config,
    entity_names: list[str] | None = None,
    metadata: dict[str, str] | None = None,
    confidence: float | None = None,
    provenance: str | None = None,
    chunk_size: int = 500,
    chunk_overlap: int = 50,
) -> ChunkResult:
    """Split content into chunks at sentence boundaries and add as linked observations."""
    result = ChunkResult()
    entity_names = entity_names or []
    metadata = metadata or {}

    chunks = _split_into_chunks(content, chunk_size, chunk_overlap)
    result.chunk_count = len(chunks)

    # Build observations from chunks
    observations = [
        ObservationInput(
            content=chunk,
            entity_names=entity_names,
            metadata={**metadata, "chunk_index": str(i), "chunk_total": str(len(chunks))},
            confidence=confidence,
            provenance=provenance,
        )
        for i, chunk in enumerate(chunks)
    ]

    # Use add_observations for the actual insertion
    add_result = await add_observations(db, observations, config)

    # Collect IDs of inserted observations
    if add_result.added > 0:
        cursor = await db.execute(
            "SELECT id FROM observations ORDER BY id DESC LIMIT ?",
            (add_result.added,),
        )
        rows = await cursor.fetchall()
        result.observation_ids = sorted(r["id"] for r in rows)

        # Set parent_id metadata on all chunks (first chunk's ID)
        if len(result.observation_ids) > 1:
            parent_id = str(result.observation_ids[0])
            placeholders = ", ".join("?" for _ in result.observation_ids)
            await db.execute(
                f"""UPDATE observations
                    SET metadata = json_set(metadata, '$.parent_id', ?)
                    WHERE id IN ({placeholders})""",
                [parent_id, *result.observation_ids],
            )
            await db.commit()

    result.errors = add_result.errors
    return result

File: src/observations.ts
import type { Backend } from "./backend";
import type { Config } from "./config";
import type { Embedder } from "./embeddings";
import { resolveEntityIds } from "./graph";
import { newId } from "./id";
import { computeStabilityGrowth } from "./scoring";
import type { ObservationInput, ObservationRow } from "./types";

export interface ObservationResult {
  id: string;
  entity_names: string[];
}

export async function addObservations(
  backend: Backend,
  observations: ObservationInput[],
  config: Config,
  embedder?: Embedder,
): Promise<ObservationResult[]> {
  if (observations.length === 0) return [];
  const results: ObservationResult[] = [];
  const ts = new Date().toISOString();

  // Batch-resolve all entity names upfront
  const allNames = observations.flatMap((o) => o.entity_names);
  const nameToId = await resolveEntityIds(backend, allNames);

  await backend.transaction(async () => {
    for (const input of observations) {
      const id = newId();
      const metadata = input.metadata ? JSON.stringify(input.metadata) : "{}";
      const confidence = input.confidence ?? 0.8;

      await backend.run(
        `INSERT INTO observations (id, content, metadata, confidence, source, created_at)
         VALUES (?, ?, ?, ?, ?, ?)`,
        [id, input.content, metadata, confidence, input.source ?? null, ts],
      );

      for (const entityName of input.entity_names) {
        await backend.run(
          "INSERT INTO observation_entities (observation_id, entity_id) VALUES (?, ?)",
          [id, nameToId.get(entityName)!],
        );
      }

      results.push({ id, entity_names: input.entity_names });
    }
  });

  // Sync FTS index for new observations
  if (results.length > 0) {
    await backend.syncObservationFts(results.map((r) => r.id));
  }

  // Bump stability on linked entities (write = active knowledge maintenance)
  const uniqueEntityIds = [...new Set(allNames)]
    .map((n) => nameToId.get(n)!)
    .filter(Boolean);
  if (uniqueEntityIds.length > 0) {
    const ph = uniqueEntityIds.map(() => "?").join(", ");
    const entities = await backend.all<{
      id: string;
      stability: number;
      last_accessed: string | null;
    }>(
      `SELECT id, stability, last_accessed FROM entities WHERE id IN (${ph})`,
      uniqueEntityIds,
    );

    // Build batched CASE expression (each entity has different stability growth)
    const ts = new Date().toISOString();
    const caseParts: string[] = [];
    const caseParams: (string | number)[] = [];
    for (const entity of entities) {
      caseParts.push("WHEN id = ? THEN ?");
      caseParams.push(
        entity.id,
        computeStabilityGrowth(entity.stability, entity.last_accessed, config),
      );
    }

    await backend.run(
      `UPDATE entities SET
        stability = CASE ${caseParts.join(" ")} END,
        last_accessed = ?
      WHERE id IN (${ph})`,
      [...caseParams, ts, ...uniqueEntityIds],
    );
  }

  // Embed after transaction succeeds (non-fatal on failure)
  if (embedder && results.length > 0) {
    try {
      const texts = observations.map((o) => o.content);
      const vectors = await embedder.embed(texts);
      await backend.storeEmbeddings(
        results.map((r, i) => ({ id: r.id, vector: vectors[i]! })),
      );
    } catch {
      // Graceful degradation: observation saved, embedding skipped
    }
  }

  return results;
}

export async function readObservation(
  backend: Backend,
  id: string,
): Promise<(ObservationRow & { entity_names: string[] }) | undefined> {
  const obs = await backend.get<ObservationRow>(
    "SELECT * FROM observations WHERE id = ?",
    [id],
  );
  if (!obs) return undefined;

  // Get linked entity names
  const entities = await backend.all<{ name: string }>(
    `SELECT e.name FROM entities e
     JOIN observation_entities oe ON oe.entity_id = e.id
     WHERE oe.observation_id = ?`,
    [id],
  );

  return {
    ...obs,
    entity_names: entities.map((e) => e.name),
  };
}

export async function addChunkedObservation(
  backend: Backend,
  content: string,
  entityNames: string[],
  config: Config,
  options?: {
    metadata?: Record<string, unknown>;
    confidence?: number;
    source?: string;
    maxChunkSize?: number;
  },
  embedder?: Embedder,
): Promise<ObservationResult[]> {
  const maxSize = options?.maxChunkSize ?? 2000;
  const chunks = splitIntoChunks(content, maxSize);
  const groupId = newId();

  const observations: ObservationInput[] = chunks.map((chunk, i) => ({
    content: chunk,
    entity_names: entityNames,
    metadata: {
      ...options?.metadata,
      chunk_index: i,
      chunk_total: chunks.length,
      chunk_group: groupId,
    },
    confidence: options?.confidence,
    source: options?.source,
  }));

  return addObservations(backend, observations, config, embedder);
}

export async function backfillEmbeddings(
  backend: Backend,
  embedder: Embedder,
  options?: { batchSize?: number; force?: boolean },
): Promise<{
  embedded: number;
  skipped: number;
  errors: number;
  last_error?: string;
}> {
  const batchSize = options?.batchSize ?? 100;
  let embedded = 0;
  let errors = 0;
  let lastError: string | undefined;

  const baseQuery = options?.force
    ? "SELECT id, content FROM observations ORDER BY created_at"
    : `SELECT o.id, o.content FROM observations o
       LEFT JOIN observation_vectors v ON v.observation_id = o.id
       WHERE v.observation_id IS NULL
       ORDER BY o.created_at`;

  const coverage = await backend.getEmbeddingCoverage();
  let skipped = options?.force ? 0 : coverage.embedded;

  // Paginated fetch to avoid loading all observations into memory
  for (let offset = 0; ; offset += batchSize) {
    const batch = await backend.all<{ id: string; content: string }>(
      `${baseQuery} LIMIT ? OFFSET ?`,
      [batchSize, offset],
    );
    if (batch.length === 0) break;

    for (let attempt = 0; attempt < 2; attempt++) {
      try {
        const texts = batch.map((r) => r.content);
        const vectors = await embedder.embed(texts);
        await backend.storeEmbeddings(
          batch.map((r, j) => ({ id: r.id, vector: vectors[j]! })),
        );
        embedded += batch.length;
        break;
      } catch (e) {
        if (attempt === 1) {
          errors += batch.length;
          lastError = e instanceof Error ? e.message : String(e);
        }
      }
    }
  }

  // Recalculate skipped for force mode (total - what we processed)
  if (options?.force) {
    skipped = coverage.total - embedded - errors;
  }

  return {
    embedded,
    skipped,
    errors,
    ...(lastError && { last_error: lastError }),
  };
}

function splitIntoChunks(text: string, maxSize: number): string[] {
  if (text.length <= maxSize) return [text];

  const chunks: string[] = [];
  let remaining = text;

  while (remaining.length > 0) {
    if (remaining.length <= maxSize) {
      chunks.push(remaining);
      break;
    }

    // Try to split at paragraph boundary
    let splitAt = remaining.lastIndexOf("\n\n", maxSize);
    if (splitAt <= 0) {
      // Try sentence boundary
      splitAt = remaining.lastIndexOf(". ", maxSize);
      if (splitAt > 0) splitAt += 2; // Include the ". "
    }
    if (splitAt <= 0) {
      // Force split at maxSize
      splitAt = maxSize;
    }

    chunks.push(remaining.slice(0, splitAt).trim());
    remaining = remaining.slice(splitAt).trim();
  }

  return chunks;
}
